pgjdbc/src/main/java/org/postgresql/jdbc/PgConnection.java

```diff
@@ -378,10 +384,170 @@ public CallableStatement prepareCall(String sql) throws SQLException {
+    return new LinkedHashMap<String, Class<?>>(getTypeMapNoCopy());
+  }
+
+  @Override
+  public Map<String, Class<?>> getTypeMapNoCopy() {
     return typemap;
   }
 
+  /**
+   * @see #getTypeMapInvertedDirect(java.lang.Class)
+   * @see #getTypeMapInvertedInherited(java.lang.Class)
+   */
+  private static void makeValuesUnmodifiable(Map<Class<?>, Set<String>> invertedMap) {
+    for (Map.Entry<Class<?>, Set<String>> entry : invertedMap.entrySet()) {
+      Set<String> types = entry.getValue();
+      if (types.size() == 1) {
+        entry.setValue(Collections.singleton(types.iterator().next()));
+      } else {
+        entry.setValue(Collections.unmodifiableSet(types));
+      }
+    }
+  }
+
+  /**
+   * The set of types directly mapped to each class - used for type inference.
+   */
+  private final Map<Class<?>, Set<String>> typemapInvertedDirect = new HashMap<Class<?>, Set<String>>();
+
+  /**
+   * The {@link #typemap} present at the time {@link #typemapInvertedDirect} was last
+   * rebuilt.
+   *
+   * @see  #getTypeMapInvertedDirect(java.lang.Class)
+   */
+  private Map<String, Class<?>> typemapInvertedDirectSource = typemap;
+
+  /**
+   * Adds a new element to an inverted map.
+   *
+   * @param invertedMap the map to add to
+   * @param clazz the class to add
+   * @param type the type that maps to this class
+   *
+   * @return  {@code true} when the class is added to the map, {@code false} when already existed.
+   *
+   * @see  #getTypeMapInvertedDirect(java.lang.Class)
+   */
+  private static boolean addInverted(Map<Class<?>, Set<String>> invertedMap, Class<?> clazz, String type) {
+    Set<String> types = invertedMap.get(clazz);
+    if (types == null) {
+      types = new HashSet<String>();
+      invertedMap.put(clazz, types);
+    }
+    boolean added = types.add(type);
+    if (added) {
+      LOGGER.log(Level.FINER, "Added: {0} -> {1}", new Object[] {clazz.getName(), type});
+    } else {
+      LOGGER.log(Level.FINEST, "Not added: {0} -> {1}", new Object[] {clazz.getName(), type});
+    }
+    return added;
+  }
+
+  @Override
+  public Set<String> getTypeMapInvertedDirect(Class<?> clazz) {
+    Set<String> types;
+    synchronized (typemapInvertedDirect) {
+      Map<String, Class<?>> source = typemap;
+      if (source != typemapInvertedDirectSource) {
+        // Recreate the inverted map when the typemap has been changed
+        typemapInvertedDirect.clear();
+        for (Map.Entry<String, Class<?>> entry : source.entrySet()) {
+          addInverted(typemapInvertedDirect, entry.getValue(), entry.getKey());
+        }
+        // Make each type set unmodifiable
+        makeValuesUnmodifiable(typemapInvertedDirect);
+        // Update source to know when to rebuild
+        typemapInvertedDirectSource = source;
+        LOGGER.log(Level.FINE, "  typemapInvertedDirect = {0}", typemapInvertedDirect);
+      }
+      types = typemapInvertedDirect.get(clazz);
+    }
+    if (types == null) {
```

Comment:
Fixed.  We had done this to avoid the funky `Collections.<String>emptySet()` syntax required for Java 6.

https://github.com/pgjdbc/pgjdbc/pull/1378/commits/ae042fac86ad45099a8472ecaa4f1495cee8a042